### PR TITLE
Rewrite flaky tests reporter into its own package

### PR DESCRIPTION
**Pull request by kevin940726**
## What?

1. Rewrite the flaky tests reporter into its own package
2. Convert it into TypeScript.
3. Add unit tests.
4. Refactor the output of the issue body.

## Why?

It's been extremely hard to add or contribute to the flaky tests reporter without tests. By converting it into TypeScript in its own package allows us to add unit tests.

## How?

Ideally, it'd be nice to just put this in `project-management-automation`, but that action is only run on `push` or `pull_request_target` event, while the reporter needs to be run after the e2e tests complete event.

I removed the estimated flaky rate since it isn't that useful in the long run. The log is now kept though so we can still generate them with a different approach in the future.

## Testing Instructions

Unit tests should pass. And https://github.com/kevin940726/gutenberg/issues/15 is an example of the result.

### CODE REVIEW

File: .github/workflows/flaky-tests.yml
Comment (noisysocks) [inline .github/workflows/flaky-tests.yml:34]: `report-flaky-tests` doesn't have an `index.js`. Is that right? How does GitHub Actions know to run `build/index.js`?

File: packages/report-flaky-tests/tsconfig.json
Comment (noisysocks) [inline packages/report-flaky-tests/tsconfig.json]: Why `lib`? In most other packages it's `src`.

File: packages/report-flaky-tests/package.json
Comment (noisysocks) [inline packages/report-flaky-tests/package.json:3]: I'm never 100% about this but I think we want this to be `1.0.0-prerelease` so that the first publish (once `"private": true` is removed) is `1.0.0`.

File: packages/report-flaky-tests/package.json
Comment (noisysocks) [inline packages/report-flaky-tests/package.json:26]: Should we add `"types": "build-types"`?

**Review by noisysocks — COMMENTED** [review]
There's a lot here 😅 It all looks like good stuff but it's difficult for me to review because GitHub shows everything as having changed. Could you split the PR or 5403c5b728161910045a4913ef2cb3b7e8e714cb up? In particular it would be helpful to move the code to a new package in a separate PR so that at least GitHub can show me which lines have changed.

File: .github/workflows/flaky-tests.yml
Comment (kevin940726) [reply .github/workflows/flaky-tests.yml:34]: I believe GH Action will look for the entry in [`action.yml`](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action#creating-an-action-metadata-file), which points to `build/index.js`.

File: packages/report-flaky-tests/tsconfig.json
Comment (kevin940726) [reply packages/report-flaky-tests/tsconfig.json]: The build steps (`npm run dev` and `npm run build`) will automatically transpile everything in `src` into `build` and `build-module` with Babel. However, it doesn't make sense to compile GH Action into ESM because that's not supported. I guess we can still use `src` though if we want to make it consistent.

File: packages/report-flaky-tests/package.json
Comment (kevin940726) [reply packages/report-flaky-tests/package.json:3]: I don't have a strong opinion about this, but I think that `1.0.0` is considered the first _stable_ release in semver, and maybe we should start with `0.x.x` for experimental/private package?

File: packages/report-flaky-tests/package.json
Comment (kevin940726) [reply packages/report-flaky-tests/package.json:26]: Oops, I forgot about that. However, I don't think we really need to output the types for this package though. It's not really designed to be consumed by consumer code, but just for the GH workflows 😅.

File: packages/report-flaky-tests/package.json
Comment (noisysocks) [reply packages/report-flaky-tests/package.json:3]: Yeah fair enough. Not sure if the project has a preference. Maybe @gziolo knows.

File: packages/report-flaky-tests/lib/index.ts
Comment (noisysocks) [inline packages/report-flaky-tests/lib/index.ts]: It probably doesn't matter in practice but a conditional is probably the safer way to coerce `error: any` into `error: Error | string`.

```ts
core.error( error instanceof Error ? error : error.toString() );
```

File: packages/report-flaky-tests/tsconfig.json
Comment (noisysocks) [reply packages/report-flaky-tests/tsconfig.json]: Do you have a preference @gziolo?

File: .github/workflows/flaky-tests.yml
Comment (noisysocks) [inline .github/workflows/flaky-tests.yml]: It's not clear why `npm run build:package-types` does what this comment says which is "build the action itself". Maybe say something like "We don't have to build the entire project, just the action itself, which uses TypeScript as its compiler."

File: packages/report-flaky-tests/lib/github-api.ts
Comment (noisysocks) [inline packages/report-flaky-tests/lib/github-api.ts]: Can we just type this as `octokit: Octokit`? I think the helpers (`ReturnType`, `typeof`) are neat but they force the reader of the code to have to think for a second which can add up.

File: packages/report-flaky-tests/lib/github-api.ts
Comment (noisysocks) [inline packages/report-flaky-tests/lib/github-api.ts]: What type does `download.data` have? Would it be better to use a guard instead of always casting to `Buffer`?

File: packages/report-flaky-tests/lib/github-api.ts
Comment (noisysocks) [inline packages/report-flaky-tests/lib/github-api.ts]: What if it's invalid JSON? Or is missing fields that have recently been added to `FlakyTestResult`?

File: packages/report-flaky-tests/lib/run.ts
Comment (noisysocks) [inline packages/report-flaky-tests/lib/run.ts]: What does this do?

File: packages/report-flaky-tests/lib/run.ts
Comment (noisysocks) [inline packages/report-flaky-tests/lib/run.ts]: Neat!

File: packages/report-flaky-tests/lib/run.ts
Comment (noisysocks) [inline packages/report-flaky-tests/lib/run.ts]: `Date` objects are comparable so the two calls to `getTime` are not necessary.

File: packages/report-flaky-tests/lib/run.ts
Comment (noisysocks) [inline packages/report-flaky-tests/lib/run.ts]: Same thing here about `err` maybe not being `Error`. Perhaps in these cases just use `as!`.

File: packages/report-flaky-tests/lib/markdown.ts
Comment (noisysocks) [inline packages/report-flaky-tests/lib/markdown.ts]: We generally have been putting all type definitions into `./types`.

File: packages/report-flaky-tests/lib/constants.ts
Comment (noisysocks) [inline packages/report-flaky-tests/lib/constants.ts]: Since they're only used in `markdown.ts` I don't think there's much value in having these constants live alone in `constants.ts`. I'd probably put it all in `markdown.ts`.

File: packages/report-flaky-tests/lib/markdown.ts
Comment (noisysocks) [inline packages/report-flaky-tests/lib/markdown.ts]: Can we add a `type` declaration for this? Just need to be careful about handling when new attributes are added.

File: packages/report-flaky-tests/lib/github-api.ts
Comment (noisysocks) [inline packages/report-flaky-tests/lib/github-api.ts]: Oh, you already did it! 😀

File: packages/report-flaky-tests/lib/github-api.ts
Comment (noisysocks) [inline packages/report-flaky-tests/lib/github-api.ts]: What's this about?

File: packages/report-flaky-tests/lib/github-api.ts
Comment (kevin940726) [reply packages/report-flaky-tests/lib/github-api.ts]: `download.data` has the type of `unknown`. I think that's an issue of the official typings of `@actions/github` and it should always be `Buffer`.

File: packages/report-flaky-tests/lib/markdown.ts
Comment (noisysocks) [inline packages/report-flaky-tests/lib/markdown.ts]: I think this would read simpler as a `for ... of` as then you wouldn't need so many `map`s or the `is` type assertion.

**Comment by noisysocks** [comment]
Very cool! I like the approach you took for adding tests. The snapshots seem correct, though you would know better than me! 😀

File: packages/report-flaky-tests/lib/run.ts
Comment (kevin940726) [reply packages/report-flaky-tests/lib/run.ts]: It simply removes the starting path of the test file:

```diff
-/home/runner/work/gutenberg/gutenberg/test/e2e/specs/editor/various/copy-cut-paste.spec.js
+/test/e2e/specs/editor/various/copy-cut-paste.spec.js
```

Only the playwright test runner will output absolute paths though, hence the condition.

File: packages/report-flaky-tests/lib/run.ts
Comment (kevin940726) [reply packages/report-flaky-tests/lib/run.ts]: Oh cool, didn't know about that! I think it under the hood just uses `.valueOf()` to convert them into numbers though.

File: packages/report-flaky-tests/lib/github-api.ts
Comment (kevin940726) [reply packages/report-flaky-tests/lib/github-api.ts]: This is super weird, I don't know if there's a better way to do this. Basically, what it does is allow all original parameters but without `repo` and `owner`.

File: packages/report-flaky-tests/tsconfig.json
Comment (kevin940726) [reply packages/report-flaky-tests/tsconfig.json]: The `lib` naming convention is copied from the `project-management-automation` package.

Comment (gziolo) [reply packages/report-flaky-tests/tsconfig.json]: In other packages we use `lib` when we don't want them to be transpiled with Babel, `@wordpress/prettier-config` and similar non-production packages would be a good example. Here, we have TypeScript files so the code won't work out of the box. That's also why `@wordpress/project-management-automation` uses JSDoc to import TypeScript types to avoid the additional step of removing TypeScript annotations from the source code.

> However, it doesn't make sense to compile GH Action into ESM because that's not supported. 

Node 14 is the minimum LTS version and it fully supports ESM. What is the reason it doesn't work with GitHub actions?

The build process we have configured produces also CommonJS version if necessary. For production packages, it's the other way around, we use only ESM and produce CommonJS for optional 3rd party usage.

File: packages/report-flaky-tests/package.json
Comment (gziolo) [reply packages/report-flaky-tests/package.json:3]: I think it doesn't matter here in case of package for development purposes.

File: .github/workflows/flaky-tests.yml
Comment (gziolo) [reply .github/workflows/flaky-tests.yml]: Or we move types to their own file and import them with JSDoc to avoid all the hassle.

**Comment by gziolo** [comment]
> Ideally, it'd be nice to just put this in project-management-automation, but that action is only run on push or pull_request_target event, while the reporter needs to be run after the e2e tests complete event.

Based on the usage

```yaml
uses: ./packages/report-flaky-tests
```

I assume it would be possible to move all the code to:

`packages/project-managemment-automation/report-flaky-tests`

or

`packages/project-managemment-automation/lib/tasks/report-flaky-tests`


You are referencing a folder on the disk so you can keep doing that if there is some limitation to what the existing `action.yml` from `@wordpress/project-management-automation` can do.

File: packages/report-flaky-tests/tsconfig.json
Comment (kevin940726) [reply packages/report-flaky-tests/tsconfig.json]: > Node 14 is the minimum LTS version and it fully supports ESM. What is the reason it doesn't work with GitHub actions?

Oh, wait. Seems like GH Actions support [node 16](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#runsusing) now. The last time I tried it only supported node 12.

> The build process we have configured produces also CommonJS version if necessary. For production packages, it's the other way around, we use only ESM and produce CommonJS for optional 3rd party usage.

For a private and internal package like this, building it into two different formats still isn't necessary, but that's a different issue. We could keep it simple and still use the default `src` and let Babel compiles it.

File: packages/report-flaky-tests/lib/markdown.ts
Comment (kevin940726) [reply packages/report-flaky-tests/lib/markdown.ts]: What do you mean? In what case would that be a problem?

File: packages/report-flaky-tests/lib/markdown.ts
Comment (kevin940726) [reply packages/report-flaky-tests/lib/markdown.ts]: I think `map`s are simpler though 😅, and `for...of` doesn't play nice with TypeScript I believe.

**Comment by kevin940726** [comment]
> Based on the usage
> 
> ```yaml
> uses: ./packages/report-flaky-tests
> ```
> 
> I assume it would be possible to move all the code to:
> 
> `packages/project-managemment-automation/report-flaky-tests`
> 
> or
> 
> `packages/project-managemment-automation/lib/tasks/report-flaky-tests`
> 
> You are referencing a folder on the disk so you can keep doing that if there is some limitation to what the existing `action.yml` from `@wordpress/project-management-automation` can do.

Wouldn't that be kind of confusing? That we have a task in the package, written in TypeScript, has its own dependencies, has its own unit tests, and is completely unrelated to the other tasks in the same directory.

File: packages/report-flaky-tests/tsconfig.json
Comment (kevin940726) [reply packages/report-flaky-tests/tsconfig.json]: I'd love to have the monorepo have individual build steps for each package defined in its `package.json`. So that we don't have to deal with this weird naming convention setting. A bonus benefit is that we can build each package separately. Like in the flaky tests reporter workflow, we don't have to build the whole project, but only the `report-flaky-tests` package. I don't think currently this can be easily done?

File: .github/workflows/flaky-tests.yml
Comment (kevin940726) [reply .github/workflows/flaky-tests.yml]: The experience of writing JSDoc is still pretty different than TypeScript, and I'd like to stick to TypeScript as much as possible for the benefits of what it brings. IMO, we should fix the setup of the project to allow such things instead of forcing developers to write JSDoc when TypeScript is available and preferred.

File: packages/report-flaky-tests/lib/markdown.ts
Comment (kevin940726) [reply packages/report-flaky-tests/lib/markdown.ts]: Not sure about this, I don't think we need to put every type in `./types` if it's only used in one file.

File: packages/report-flaky-tests/tsconfig.json
Comment (noisysocks) [reply packages/report-flaky-tests/tsconfig.json]: Think I'm leaning towards `src` just because `lib` implies that no compilation is required whatsoever which isn't true here (it requires `tsc`).

> I don't think currently this can be easily done?

I suppose you you could add a script to the `package.json` in `report-flaky-tests` and `cd report-flaky-tests; npm run ...`? It's a bit obtuse, I probably wouldn't recommend it.

File: .github/workflows/flaky-tests.yml
Comment (noisysocks) [reply .github/workflows/flaky-tests.yml]: I also think TypeScript is a good fit here as it adds safety to something that's error-prone but difficult to test on your local machine.

File: packages/report-flaky-tests/lib/github-api.ts
Comment (noisysocks) [reply packages/report-flaky-tests/lib/github-api.ts]: Cool 👍 thanks for checking

File: packages/report-flaky-tests/lib/markdown.ts
Comment (noisysocks) [reply packages/report-flaky-tests/lib/markdown.ts]: It's the convention that we've been using in `@wordpress/components` which is where most of the project's TypeScript is currently. I think the motivation is to make it so that there is less "TypeScript noise" to read while reading the code. It's not a big deal either way.

File: packages/report-flaky-tests/lib/markdown.ts
Comment (noisysocks) [reply packages/report-flaky-tests/lib/markdown.ts]: Sorry, so there's two comments here.

> Can we add a `type` declaration for this?

By this I just mean can we add a `type` alias for the `{ }` type that is returned by this function. Giving complex types a name improves readability in the same way that using variables improves readability. In particular I'm keen on anything that reduces the amount of "TypeScript noise" that can be distracting when reading code.

> Just need to be careful about handling when new attributes are added.

By this I mean we need to be careful that if, in the future, a new attribute is added to this type, that it is marked optional, because existing data that exists in GitHub will not have that new attribute. I'm not sure if there's a nice way to inform future maintainers of that—probably there's nothing we can do.

File: packages/report-flaky-tests/tsconfig.json
Comment (kevin940726) [reply packages/report-flaky-tests/tsconfig.json]: > Think I'm leaning towards `src` just because `lib` implies that no compilation is required whatsoever which isn't true here (it requires `tsc`).

I've changed it to `src` in https://github.com/WordPress/gutenberg/pull/41106/commits/9c451ee01860b5299ff73258df37804e77f5a104 👍

> I suppose you you could add a script to the `package.json` in `report-flaky-tests` and `cd report-flaky-tests; npm run ...`? It's a bit obtuse, I probably wouldn't recommend it.

I guess that could work? But then I _think_ we have to also move some dev dependencies there (`babel` or `tsc` for instance). Either way, the built result will still be in the `build` folder, which conflicts with the default result when the code is in `src`. I guess that would be a little weird 😅.

File: packages/report-flaky-tests/lib/github-api.ts
Comment (noisysocks) [reply packages/report-flaky-tests/lib/github-api.ts]: What is the `& { issue_number }` for?

Why can't we have `owner`?

I think it would be clearer and more valuable to use the type that is accepted by `octokit.rest.issues.update` which is `RestEndpointMethodTypes["issues"]["update"]["parameters"]` instead of using the `Parameters<>` helper. 

Right now this type doesn't offer much value because it's just restating the code that is below in `updateIssue` but in a different way. It's a bit like when you see a comment like this:

```js
// Print foo.
print( foo );
```

Ideally the type is a way in which we can express what our expectations are about the function. A bit like a unit test. The compiler then aids us by ensuring that the code matches our expectations. 

```ts
type UpdateIssueParams = Omit<
		RestEndpointMethodTypes["issues"]["update"]["parameters"],
		'repo'
>;
async updateIssue( params: UpdateIssueParams ) {
}
```

File: packages/report-flaky-tests/tsconfig.json
Comment (noisysocks) [reply packages/report-flaky-tests/tsconfig.json]: I think it could get confusing, let's stick to having scripts at the top level only.

File: packages/report-flaky-tests/lib/markdown.ts
Comment (noisysocks) [reply packages/report-flaky-tests/lib/markdown.ts]: OK, let's call it developer preference!

But at least get rid of the `ReturnType< typeof parseFormattedTestResults >`. If you add a new `type` declaration—e.g. `TestResult`—for the return type of `parseFormattedTestResults` then you can write `TestResult` instead of `ReturnType< typeof parseFormattedTestResults >`. This is easier to read and offers more value because we're asserting something useful instead of writing the same code twice: once in the type declaration and once in the body.

File: packages/report-flaky-tests/lib/markdown.ts
Comment (kevin940726) [reply packages/report-flaky-tests/lib/markdown.ts]: I don't think that's _noise_ though. Types should also be seen as documentation, just like JSDoc. If it's noise, then we should probably do the same with inline JSDoc and move most of them to a file like `jsdoc.js`. 😅

Another disadvantage of moving standalone types to `types.ts` is that we no longer know where the types are imported at first glance. It creates a mental overhead of not being able to relate the types with the file together.

I don't feel strongly about this either though, so if anyone thinks it should be consistent then I'd be happy to move it :).

File: packages/report-flaky-tests/lib/markdown.ts
Comment (kevin940726) [reply packages/report-flaky-tests/lib/markdown.ts]: Oh, it's the same type with `ParsedTestResult`, I totally forgot to change that 😅.

File: packages/report-flaky-tests/lib/github-api.ts
Comment (kevin940726) [reply packages/report-flaky-tests/lib/github-api.ts]: > Why can't we have `owner`?

Because that's passed in the function body automatically.

Cool, didn't know that `@octokit/types` is available! I'll use it as it clearly makes the types easier to follow.

My only opinion on this is that I don't think assigning the type to a different name necessarily makes it easier to read and follow. If I hover on the `updateIssue` function in another file, I don't really want to see an arbitrary `UpdateIssueParams` name in the type, but rather prefer the resolved type showing what exactly it accepts. There's a balance to make though, as we don't want verbose types to be resolved recursively either.

Let me update the types and see if it's easy enough to follow.

File: packages/report-flaky-tests/lib/markdown.ts
Comment (kevin940726) [reply packages/report-flaky-tests/lib/markdown.ts]: True! Agree about that 👍

File: packages/report-flaky-tests/lib/github-api.ts
Comment (kevin940726) [reply packages/report-flaky-tests/lib/github-api.ts]: > What if it's invalid JSON?

Then I believe the whole action would just fail. I don't think we can do anything about that? Just to make sure the artifact being uploaded has valid JSON.

> Or is missing fields that have recently been added to `FlakyTestResult`?

Good question! Not sure about that though. The action will always run on `trunk` so I don't think we have to worry about new fields? But there might be a problem with outdated artifacts though. I guess we just have to make sure it's backward compatible, or introduce something like `version` field. Open to ideas!

Comment (noisysocks) [reply packages/report-flaky-tests/lib/github-api.ts]: A `version` field could be good. If nothing else it might encourage future developers to consider this problem before adding or changing a field.

File: packages/report-flaky-tests/lib/markdown.ts
Comment (noisysocks) [reply packages/report-flaky-tests/lib/markdown.ts]: >  If it's noise, then we should probably do the same with inline JSDoc and move most of them to a file like jsdoc.js. 😅

We sometimes do in a way, when we do the `/** @typedef Blah import( './Blah' ) */` thing 😀

But all good, I don't feel strongly about this.

File: packages/report-flaky-tests/lib/github-api.ts
Comment (noisysocks) [reply packages/report-flaky-tests/lib/github-api.ts]: Yeah that's fair enough. Mostly I just hate using lots of helper types like `Parameters<>`, `Omit<>`, etc. because it means that you really have to think when hovering over a function and seeing the type. I don't ever want to think! 😛

**Review by noisysocks — APPROVED** [review]
This is looking good, thanks for making those changes.

I think it's worth considering adding a `version` field to the metadata / artifacts for forwards compatibility and as a subtle hint to future developers that existing data is something that they need to consider when making changes. Up to you, though.

It's obviously difficult to test this but the code looks reasonable, the unit tests you've added are passing, and the typings you've added are compiling. I've restarted the failed E2E tests which seem unrelated. When you're ready to merge this, please do it in your morning so that you're around in case there are any problems 🙏

**Comment by noisysocks** [comment]
PS. I think it would be good to discuss the flaky test reporter with the wider Gutenberg team e.g. in a #core-editor meeting, in a GitHub discussion, or in a GitHub issue. I see that most issues it creates go unnoticed which is either a sign that it's not useful enough for devs, or that nobody knows about it, or that we are not prioritising fixing flaky tests. Probably all three! I know you were thinking about building a web dashboard for flaky tests instead of using Github Issues—I think that's a neat idea and also worth discussing.

**Comment by kevin940726** [comment]
> PS. I think it would be good to discuss the flaky test reporter with the wider Gutenberg team e.g. in a #core-editor meeting, in a GitHub discussion, or in a GitHub issue. I see that most issues it creates go unnoticed which is either a sign that it's not useful enough for devs, or that nobody knows about it, or that we are not prioritising fixing flaky tests. Probably all three! I know you were thinking about building a web dashboard for flaky tests instead of using Github Issues—I think that's a neat idea and also worth discussing.

Yeah, probably all three 😅.

FWIW, the flaky test reports originally started with the motivation of re-running failed e2e tests in PR to unblock developers and contributors. The issues are only by-products of the results to keep track of the flaky tests that went unnoticed.

GH issues are probably not the best way to record these data though, so I think it's totally expected that some of them never got the attention they deserve. Improving the reporter (like this PR) is an attempt to try to make it more useful, and potentially pave the way to transform it into something essential to the development process.

We can discuss further ideas with the community, but this is still pretty much an internal tool for project management, so I think keeping the discussion small and focused might be better for the overall process.

File: packages/report-flaky-tests/lib/github-api.ts
Comment (kevin940726) [reply packages/report-flaky-tests/lib/github-api.ts]: Added a `version` field to the reports!

**Pull request merged**
